### PR TITLE
Refs #34597 -- Avoided unnecessary constraints on reused split_exclude.

### PR DESCRIPTION
**Pull request by charettes**
Addresses a SQL regression introduced by [8593e162c9cb63a6c0b06daf045bc1c21eb4d7c1](https://code.djangoproject.com/changeset/8593e162c9cb63a6c0b06daf045bc1c21eb4d7c1/) that is detailed in ticket-34597.

The SQL diff of this change is

```diff
--- test_ticket14511 (queries.tests.ExcludeTests):sqlite:89f10a80d7
+++ test_ticket14511 (queries.tests.ExcludeTests):sqlite:500a2087c2
@@ -161,7 +161,6 @@
                   (SELECT %s AS "a"
                    FROM "queries_employment" U1
                    WHERE (U1."title" IN (%s, %s)
-                          AND U1."id" = ("queries_employment"."id")
-                          AND "queries_employment"."employer_id" = ("queries_company"."id"))
+                          AND U1."id" = ("queries_employment"."id"))
                    LIMIT 1)))
 ORDER BY "queries_company"."name" ASC
--- test_ticket_24605 (queries.tests.TestTicket24605):sqlite:89f10a80d7
+++ test_ticket_24605 (queries.tests.TestTicket24605):sqlite:500a2087c2
@@ -25,7 +25,6 @@
               FROM "Individual" U0
               LEFT OUTER JOIN "RelatedIndividual" U1 ON (U0."id" = U1."related_id")
               WHERE (U1."id" IS NULL
-                     AND U0."id" = ("Individual"."id")
-                     AND "Individual"."id" = ("Individual"."id"))
+                     AND U0."id" = ("Individual"."id"))
               LIMIT 1))
 ORDER BY "Individual"."id" ASC
--- test_ticket_23605 (queries.tests.Ticket23605Tests):sqlite:89f10a80d7
+++ test_ticket_23605 (queries.tests.Ticket23605Tests):sqlite:500a2087c2
@@ -22,7 +22,7 @@
      FROM "queries_ticket23605a" W0
      INNER JOIN "queries_ticket23605b" W1 ON (W0."ticket23605aparent_ptr_id" = W1."modela_fk_id")
      INNER JOIN "queries_ticket23605c" W2 ON (W1."modelc_fk_id" = W2."id")
-     INNER JOIN "queries_ticket23605b" W4 ON (W0."ticket23605aparent_ptr_id" = W4."modela_fk_id")
+     INNER JOIN "queries_ticket23605b" W3 ON (W0."ticket23605aparent_ptr_id" = W3."modela_fk_id")
      WHERE (W1."field_b0" >= (%s / W2."field_c0")
             AND W1."field_b1"
             AND NOT (EXISTS
@@ -35,10 +35,9 @@
                                   WHERE NOT (U0."field_b1"
                                              AND U0."field_b0" >= (%s / U1."field_c0")
                                              AND U0."field_b0" IS NOT NULL))
-                               AND V1."id" = (W1."id")
-                               AND W1."modela_fk_id" = (W0."ticket23605aparent_ptr_id"))
+                               AND V1."id" = (W1."id"))
                         LIMIT 1))
-            AND W4."field_b1"))
+            AND W3."field_b1"))
 SELECT "queries_ticket23605aparent"."id",
        "queries_ticket23605a"."ticket23605aparent_ptr_id"
 FROM "queries_ticket23605a"
@@ -48,7 +47,7 @@
               FROM "queries_ticket23605a" W0
               INNER JOIN "queries_ticket23605b" W1 ON (W0."ticket23605aparent_ptr_id" = W1."modela_fk_id")
               INNER JOIN "queries_ticket23605c" W2 ON (W1."modelc_fk_id" = W2."id")
-              INNER JOIN "queries_ticket23605b" W4 ON (W0."ticket23605aparent_ptr_id" = W4."modela_fk_id")
+              INNER JOIN "queries_ticket23605b" W3 ON (W0."ticket23605aparent_ptr_id" = W3."modela_fk_id")
               WHERE (W1."field_b0" >= (%s / W2."field_c0")
                      AND W1."field_b1"
                      AND NOT (EXISTS
@@ -61,7 +60,6 @@
                                            WHERE NOT (U0."field_b1"
                                                       AND U0."field_b0" >= (%s / U1."field_c0")
                                                       AND U0."field_b0" IS NOT NULL))
-                                        AND V1."id" = (W1."id")
-                                        AND W1."modela_fk_id" = (W0."ticket23605aparent_ptr_id"))
+                                        AND V1."id" = (W1."id"))
                                  LIMIT 1))
-                     AND W4."field_b1")))
+                     AND W3."field_b1")))
```

### CODE REVIEW

File: django/db/models/sql/query.py [hunk 1]
```diff
@@ -1996,10 +1996,11 @@ def split_exclude(self, filter_expr, can_reuse, names_with_path):
             lookup = lookup_class(pk.get_col(query.select[0].alias), pk.get_col(alias))
             query.where.add(lookup, AND)
             query.external_aliases[alias] = True
+        else:
+            lookup_class = select_field.get_lookup("exact")
+            lookup = lookup_class(col, ResolvedOuterRef(trimmed_prefix))
+            query.where.add(lookup, AND)
```
Comment (shangxiao) [inline django/db/models/sql/query.py:2002]: When I initially stepped into this function wrt this ticket I was also thinking of excluding this particular block so good to see my assumption confirmed 👍

**Review by shangxiao — APPROVED** [review]

File: django/db/models/sql/query.py [hunk 2]
```diff
@@ -1994,7 +1994,12 @@ def split_exclude(self, filter_expr, can_reuse, names_with_path):
             filter_rhs = OuterRef(filter_rhs)
         elif isinstance(filter_rhs, F):
             filter_rhs = OuterRef(filter_rhs.name)
-        query.add_filter(filter_lhs, filter_rhs)
+        negate = False
+        if filter_rhs is None:
```
Comment (charettes) [inline django/db/models/sql/query.py:1998]: I guess we could also handle `__isnull=True` as well

File: django/db/models/sql/query.py [hunk 3]
```diff
@@ -2014,11 +2019,15 @@ def split_exclude(self, filter_expr, can_reuse, names_with_path):
...
+            lookup = lookup_class(col, ResolvedOuterRef(trimmed_prefix))
+            query.where.add(lookup, AND)
 
-        lookup_class = select_field.get_lookup("exact")
-        lookup = lookup_class(col, ResolvedOuterRef(trimmed_prefix))
-        query.where.add(lookup, AND)
-        condition, needed_inner = self.build_filter(Exists(query))
+        exists = Exists(query)
+        if negate:
+            exists = ~Q(exists)
```
Comment (kezabelle) [inline django/db/models/sql/query.py:2029]: Does this _need_ to be wrapped into a `Q` (i.e. for API reasons that I can't "_see_" in the diff)? Off the top of my head, I _feel_ like `~Exists(query)` ... exists (though I may be mis-remembering) and could be something like:

```suggestion
        if negate:
            exists = ~exists
```

Comment (charettes) [reply django/db/models/sql/query.py:2029]: I explicitly used `~Q(exists)` because the last commit is able to reduce `~Q(~Q(foo)))` to `foo` while `~Exists` is turned into `~Q(Exact(NegatedExpression(Exists(...)), True))` which is not reducable and thus generates SQL that is harder to interpret when debugging this problem.

File: django/db/models/sql/where.py [hunk 4]
```diff
@@ -118,6 +118,17 @@ def as_sql(self, compiler, connection):
...
+            isinstance(child, WhereNode) and child.negated for child in self.children
+        ):
+            return self.create(
+                [
+                    child.create(child.children, child.connector)
+                    for child in self.children
+                ],
+                self.connector,
+            ).as_sql(compiler, connection)
+
```
Comment (shangxiao) [inline django/db/models/sql/where.py:131]: Commit message has "NOT (NOT A0 ... NOT AN) can be simplified to A0 ... NOT AN"

Isn't de morgan basically flipping and/or and !/(not!).

So "NOT (NOT A0 AND ... NOT AN) <=> (A0 OR ... AN)" 🤔

File: django/db/models/sql/query.py [hunk 5]
```diff
@@ -1994,7 +1994,12 @@ def split_exclude(self, filter_expr, can_reuse, names_with_path):
             filter_rhs = OuterRef(filter_rhs)
         elif isinstance(filter_rhs, F):
             filter_rhs = OuterRef(filter_rhs.name)
-        query.add_filter(filter_lhs, filter_rhs)
+        negate = False
+        if filter_rhs is None:
+            negate = True
+            query.add_filter(f"{filter_lhs}__isnull", False)
```
Comment (shangxiao) [inline django/db/models/sql/query.py:2000]: Stupid question time: why is this needed? Seems like it's adding a redundant filter in the where:

```
SELECT 1 AS "a"
FROM "ticket_34597_expensive_exists_translation" u1
WHERE (u1."id" IS NOT NULL
...
```

File: django/db/models/sql/where.py (hunk 4 above)
Comment (shangxiao) [reply django/db/models/sql/where.py:131]: Just an observation: the processing path is slightly different for `all()`'s vacuous truth, ie `self.negated` is true but `self.children` is empty.

I _think_ the output is the same but it's always worth double-checking.

File: django/db/models/sql/where.py [hunk 6]
```diff
@@ -118,6 +118,17 @@ def as_sql(self, compiler, connection):
...
         """
+        if self.negated and all(
+            isinstance(child, WhereNode) and child.negated for child in self.children
+        ):
+            return self.create(
+                [
+                    child.create(child.children, child.connector)
+                    for child in self.children
+                ],
+                self.connector,
```
Comment (shangxiao) [inline django/db/models/sql/where.py:129]:
```suggestion
                AND if self.connector == OR else AND,
```

Actually yep I think we need to flip this.  Also we can't branch here if `self.connector` is XOR (there is no de morgan for xor is there? 🤔

Comment (shangxiao) [reply django/db/models/sql/where.py:129]: Also couple of more things:

- Would be good to add a test specifically for this mini-featurette
- A comment might be nice for future maintainers that don't recognise the pattern that we're applying de morgan

**Comment by charettes** [comment]
Thanks to both of your review so far! I didn't get a change to look at this during the weekend but I should have time this week 🙇

File: django/db/models/sql/where.py (hunk 6 above)
Comment (charettes) [reply django/db/models/sql/where.py:130]: You are completely right @shangxiao, I'll make sure to add tests for this particular change and make the propose adjustments.

**Comment by charettes** [comment]
Alright so I break the PR in four commits

1. The first one addresses a simple comment that was not kept up-to-date
2. The second one addresses the superflous `foo = foo` lookups that were being added in the exists clause and were identified by @shangxiao as potentially causing materialization

The SQL diff for these two commits is

```
--- test_ticket14511 (queries.tests.ExcludeTests):sqlite:89f10a80d7
+++ test_ticket14511 (queries.tests.ExcludeTests):sqlite:500a2087c2
@@ -161,7 +161,6 @@
                   (SELECT %s AS "a"
                    FROM "queries_employment" U1
                    WHERE (U1."title" IN (%s, %s)
-                          AND U1."id" = ("queries_employment"."id")
-                          AND "queries_employment"."employer_id" = ("queries_company"."id"))
+                          AND U1."id" = ("queries_employment"."id"))
                    LIMIT 1)))
 ORDER BY "queries_company"."name" ASC
--- test_ticket_24605 (queries.tests.TestTicket24605):sqlite:89f10a80d7
+++ test_ticket_24605 (queries.tests.TestTicket24605):sqlite:500a2087c2
@@ -25,7 +25,6 @@
               FROM "Individual" U0
               LEFT OUTER JOIN "RelatedIndividual" U1 ON (U0."id" = U1."related_id")
               WHERE (U1."id" IS NULL
-                     AND U0."id" = ("Individual"."id")
-                     AND "Individual"."id" = ("Individual"."id"))
+                     AND U0."id" = ("Individual"."id"))
               LIMIT 1))
 ORDER BY "Individual"."id" ASC
--- test_ticket_23605 (queries.tests.Ticket23605Tests):sqlite:89f10a80d7
+++ test_ticket_23605 (queries.tests.Ticket23605Tests):sqlite:500a2087c2
@@ -22,7 +22,7 @@
      FROM "queries_ticket23605a" W0
      INNER JOIN "queries_ticket23605b" W1 ON (W0."ticket23605aparent_ptr_id" = W1."modela_fk_id")
      INNER JOIN "queries_ticket23605c" W2 ON (W1."modelc_fk_id" = W2."id")
-     INNER JOIN "queries_ticket23605b" W4 ON (W0."ticket23605aparent_ptr_id" = W4."modela_fk_id")
+     INNER JOIN "queries_ticket23605b" W3 ON (W0."ticket23605aparent_ptr_id" = W3."modela_fk_id")
      WHERE (W1."field_b0" >= (%s / W2."field_c0")
             AND W1."field_b1"
             AND NOT (EXISTS
@@ -35,10 +35,9 @@
                                   WHERE NOT (U0."field_b1"
                                              AND U0."field_b0" >= (%s / U1."field_c0")
                                              AND U0."field_b0" IS NOT NULL))
-                               AND V1."id" = (W1."id")
-                               AND W1."modela_fk_id" = (W0."ticket23605aparent_ptr_id"))
+                               AND V1."id" = (W1."id"))
                         LIMIT 1))
-            AND W4."field_b1"))
+            AND W3."field_b1"))
 SELECT "queries_ticket23605aparent"."id",
        "queries_ticket23605a"."ticket23605aparent_ptr_id"
 FROM "queries_ticket23605a"
@@ -48,7 +47,7 @@
               FROM "queries_ticket23605a" W0
               INNER JOIN "queries_ticket23605b" W1 ON (W0."ticket23605aparent_ptr_id" = W1."modela_fk_id")
               INNER JOIN "queries_ticket23605c" W2 ON (W1."modelc_fk_id" = W2."id")
-              INNER JOIN "queries_ticket23605b" W4 ON (W0."ticket23605aparent_ptr_id" = W4."modela_fk_id")
+              INNER JOIN "queries_ticket23605b" W3 ON (W0."ticket23605aparent_ptr_id" = W3."modela_fk_id")
               WHERE (W1."field_b0" >= (%s / W2."field_c0")
                      AND W1."field_b1"
                      AND NOT (EXISTS
@@ -61,7 +60,6 @@
                                            WHERE NOT (U0."field_b1"
                                                       AND U0."field_b0" >= (%s / U1."field_c0")
                                                       AND U0."field_b0" IS NOT NULL))
-                                        AND V1."id" = (W1."id")
-                                        AND W1."modela_fk_id" = (W0."ticket23605aparent_ptr_id"))
+                                        AND V1."id" = (W1."id"))
                                  LIMIT 1))
-                     AND W4."field_b1")))
+                     AND W3."field_b1")))
```

If necessary I can add extra assertion to existing tests to ensure proper SQL is generated.

As for the third and fourth ones the optimization is not as trivial as I had initially thought. Even if `exclude(m2m=None)` can be optimized to `EXISTS(m2m rows)` (which is the reported use case) things get much more complicated when dealing with `exclude(m2m_possibly_null_field=None)` and the two scenarios are hard to differentiate within the limited scope `split_exclude` receives. A larger refactor would be needed to get it to work which goes beyond the scope of this ticket until we hear back from the reporter.

If merging 1. and 2. seems acceptable for now I can close this MR and open a new one only containing these two changes to keep the discussion in this PR focused on the two other commits which be relevant in the future.

**Comment by felixxm** [comment]
@charettes Thanks :+1: I pushed the first two commits to #16974 with a small assertion that should help avoid future regressions.

**Comment by charettes** [comment]
@felixxm thanks, I'll close this PR for now then as there are other areas I'm already focusing on.
